File: python/data_structures/avl.py
"""custom implementation of a self balanced binary search tree with the purpose of practice
the ADT contains the following methods:
    - insert
    - search
    - travese
    - delete
    - successor
"""
from data_structures import bst

class AVL(bst.BST):
    class Node():
        """Node basic chainable storage unit
        """
        def __init__(self, x=None):
            self.value = x
            self.height = 1
            self.left = None
            self.right = None
        def __repr__(self):
            return repr(self.value)

    def __init__(self, x=[]):
        self.root = None
        for e in x:
            self.insert(e)

    def __repr__(self):
        res = []
        def r(node, level=0):
            if not node: return
            r(node.left, level+1)
            res.append('\t'*level + f'-->({node.value}){node.height}')
            r(node.right, level+1)
        r(self.root)
        return '\n'.join(res)

    def __len__(self):
        """depth of the tree
        """
        return self.root.height

    def rotate_left(self, pivot):
        """left rotation...
        https://en.wikipedia.org/wiki/AVL_tree#Simple_rotation
        """
        aux_node = self.Node(pivot.value)
        aux_node.left, aux_node.right = pivot.left, pivot.right.left
        # replace
        pivot.value = pivot.right.value
        pivot.left, pivot.right = aux_node, pivot.right.right
        # update weights
        self.update_weights(pivot.left)
        self.update_weights(pivot)

    def rotate_right(self, pivot):
        """right rotation...
        https://en.wikipedia.org/wiki/AVL_tree#Simple_rotation
        """
        aux_node = self.Node(pivot.value)
        aux_node.left, aux_node.right = pivot.left.right, pivot.right
        # replace
        pivot.value = pivot.left.value
        pivot.left, pivot.right = pivot.left.left, aux_node
        # update weights
        self.update_weights(pivot.right)
        self.update_weights(pivot)

    def update_weights(self, node):
        """method to compute the weights of a node given the weights of its children
        the default weight of a node is 1, and the null node is 0
        """
        height = lambda x: x.height if x else 0
        node.height = max(height(node.left), height(node.right)) + 1

    def insert(self, x):
        """insert a new node into the bst, and rebalance
        """
        # special case: empty bst
        if not self.root: self.root = self.Node(x); return
        # execute normal insertion
        def r(current_node, x):
            if current_node.value == x: raise ValueError('No duplicates allowed!')
            if x < current_node.value:
                if current_node.left:
                    r(current_node.left, x)
                else:
                    current_node.left = self.Node(x)
            else:
                if current_node.right:
                    r(current_node.right, x)
                else:
                    current_node.right = self.Node(x)
            # update the weights of each visited node
            self.update_weights(current_node)
            # rebalance violations in balance_factor with rotations...
            height = lambda x: x.height if x else 0
            balance_factor = height(current_node.right) - height(current_node.left)
            if balance_factor < -1:  # case: left
                if x > current_node.left.value:  # case: left-right
                    self.rotate_left(current_node.left)
                self.rotate_right(current_node)
                return
            if balance_factor > 1:  #case: right
                if x < current_node.right.value:  # case: right-left
                    self.rotate_right(current_node.right)
                self.rotate_left(current_node)
                return
        r(self.root, x)

    def delete(self, value):
        """search for value, and rotate until leaf, then delete
        """
        if value == None or not self.root: raise KeyError(f'{value} not found')
        # standard binary search but keep the parent
        def get(node, parent):
            if not node: raise KeyError(f'{value} not found')
            if node.value == value: return (node, parent)
            if value < node.value: return get(node.left, node)
            else: return get(node.right, node)
        node, parent = get(self.root, None)
        # bring it down to leaf... then is trivial to erase
        def r(node, parent):
            # are you root?
            if not node.left and not node.right:
                if not parent: self.root = None
                if node == parent.right: parent.right = None; return
                else: parent.left = None; return
            # rotate until leaf...
            child = node.left or node.right
            if child == node.left:
                self.rotate_right(node); r(node.right, node)
            else:
                self.rotate_left(node); r(node.left, node)
            # update weights...
            self.update_weights(node)
        r(node, parent)

    def search(self, value):
        """standard search for 'x' in the tree
        """
        if not self.root: raise ValueError(f'{value} not found')
        def r(node):
            if not node: raise ValueError(f'{value} not found')
            if value == node.value: return node
            if value < node.value: return r(node.left)
            else: return r(node.right)
        return r(self.root)

File: cpp/src/avl/src/avl.cpp
#include "avl/avl.h"

namespace avl
{

template<typename T>
Node<T>::Node(T _value): value(_value) {}

template<typename T>
AVL<T>::AVL() {}

template<typename T>
AVL<T>::AVL(const std::vector<T>& data) {
	for(T e: data)
		this->insert(e);
}

template<typename T>
void AVL<T>::insert(const T& key) {
	if(!this->root) {
		this->root = std::shared_ptr<Node<T>>(new Node<T>(key));
		return;
	}
	return this->insert(key, this->root);
}

template<typename T>
int AVL<T>::height(std::shared_ptr<Node<T>> node){
	if(!node)
		return 0;
	int h_left = 0, h_right = 0;
	if(node->left)
		h_left += node->left->height;
	if(node->right)
		h_right += node->right->height;
	return std::max(h_left, h_right) + 1;
}

template<typename T>
void AVL<T>::insert(const T& key, std::shared_ptr<Node<T>> node) {
	if(!node)
		return;
	//standard insertion
	if(node->value == key)
		throw std::invalid_argument("No, duplicates allowed!");
	if(key < node->value) {
		if(node->left)
			this->insert(key, node->left);
		else
			node->left = std::shared_ptr<Node<T>>(new Node<T>(key));
	}
	else {
		if(node->right)
			this->insert(key, node->right);
		else
			node->right = std::shared_ptr<Node<T>>(new Node<T>(key));
	}
	//update height
	node->height = this->height(node);
	//repair violations
	auto balance_factor = this->height(node->right) - this->height(node->left);
	//right rotation
	if(balance_factor < -1) {
		//compound left rotation
		if(key > node->left->value)
			this->rotate_left(node->left);
		this->rotate_right(node);
	}
	//left rotation
	else if(balance_factor > 1){
		//compound right rotation
		if(key < node->right->value)
			this->rotate_right(node->right);
		this->rotate_left(node);
	}
}

template<typename T>
std::shared_ptr<Node<T>> AVL<T>::max() {
	if(!this->root)
		return nullptr;
	auto node = this->root;
	while(node->right)
		node = node->right;
	return node;
}

template<typename T>
std::shared_ptr<Node<T>> AVL<T>::min() {
	if(!this->root)
		return nullptr;
	auto node = this->root;
	while(node->left)
		node = node->left;
	return node;
}

template<typename T>
std::shared_ptr<Node<T>> AVL<T>::successor(const T& key) {
	if(!this->root)
		return nullptr;
	return this->successor(key, this->root);
}

template<typename T>
std::shared_ptr<Node<T>> AVL<T>::successor(const T& key, 
										   std::shared_ptr<Node<T>> node,
										   std::shared_ptr<Node<T>> ancestor) {
	if(!node)
		return ancestor;
	if(node->value == key) {
		//no right child to explore, return the next biggest ancestor
		if(!node->right)
			return ancestor;
		//explore and return the min of the left subtree
		else {
			auto tmp = node->right;
			while(tmp->left)
				tmp = tmp->left;
			return tmp;
		}
	}
	if(key < node->value)
		return this->successor(key, node->left, node);
	else
		return this->successor(key, node->right, ancestor);
}

/// standard binary search
template<typename T>
std::shared_ptr<Node<T>> AVL<T>::search(const T& key) {
	if(!this->root)
		return nullptr;
	return this->search(key, this->root);
}

template<typename T>
std::shared_ptr<Node<T>> AVL<T>::search(const T& key,
										std::shared_ptr<Node<T>> node) {
	if(!node)
		return nullptr;
	if(key == node->value)
		return node;
	if(key < node->value)
		return this->search(key, node->left);
	else
		return this->search(key, node->right);
}

/// standard right rotation: https://en.wikipedia.org/wiki/AVL_tree#Simple_rotation
template<typename T>
void AVL<T>::rotate_right(std::shared_ptr<Node<T>> node) {
	if(!node)
		return;
	//set-up
	auto tmp = std::shared_ptr<Node<T>>(new Node<T>(node->value));
	tmp->left = node->left->right;
	tmp->right = node->right;
	//rotate
	node->value = node->left->value;
	node->left = node->left->left;
	node->right = tmp;
	//update height...
	node->right->height = this->height(node->right);
	node->height = this->height(node);
}

/// standard left rotation: https://en.wikipedia.org/wiki/AVL_tree#Simple_rotation
template<typename T>
void AVL<T>::rotate_left(std::shared_ptr<Node<T>> node) {
	if(!node)
		return;
	//set-up
	auto tmp = std::shared_ptr<Node<T>>(new Node<T>(node->value));
	tmp->left = node->left;
	tmp->right = node->right->left;
	//rotate
	node->value = node->right->value;
	node->left = tmp;
	node->right = node->right->right;
	//update height...
	node->left->height = this->height(node->left);
	node->height = this->height(node);
}

/// initializer for the recursive erase function
template<typename T>
void AVL<T>::erase(const T& key) {
	if(!this->root)
		return;
	std::shared_ptr<Node<T>> father = nullptr;
	return erase(key, this->root, father);
}

/// rotate the node to be erased until it becomes a leaf and then delete is trivial
template<typename T>
void AVL<T>::erase(const T& key,
				   std::shared_ptr<Node<T>> node,
				   std::shared_ptr<Node<T>> father) {
	if(!node)
		return;
	if(node->value == key) {
		//have children? keep rotating...
		if(node->left || node->right) {
			auto child = (node->left)?node->left:node->right;
			if(node->left == child) {
				this->rotate_right(node);
				this->erase(key, node->right, node);
			}
			else {
				this->rotate_left(node);
				this->erase(key, node->left, node);
			}
		}
		//no children => erase the leaf easily
		else {
			//special case: are you root?
			if(!father){
				this->root = nullptr;
				return;
			}
			//general case:
			if(father->left == node)
				father->left = nullptr;
			else
				father->right = nullptr;
			return;
		}
	}
	//keep looking
	else {
		if(key < node->value)
			this->erase(key, node->left, node);
		else
			this->erase(key, node->right, node);
	}
	//update height
	node->height = this->height(node);
	//repair violations
	auto balance_factor = this->height(node->right) - this->height(node->left);
	//right rotation
	if(balance_factor < -1) {
		//compound left rotation
		if(this->height(node->left->right) > this->height(node->left->left))
			this->rotate_left(node->left);
		this->rotate_right(node);
	}
	//left rotation
	else if(balance_factor > 1){
		//compound right rotation
		if(this->height(node->right->left) > this->height(node->right->right))
			this->rotate_right(node->right);
		this->rotate_left(node);
	}
}

template<typename T>
void AVL<T>::traverse() {
	if(!this->root)
		return;
	return traverse(this->root);
}

template<typename T>
void AVL<T>::traverse(std::shared_ptr<Node<T>> node, int level) {
	if(!node)
		return;
	this->traverse(node->left, level+1);
	for(auto i = 0; i < level; i++)
		std::cout << "\t";
	std::cout << "-->(" << std::to_string(node->value) << ")"
				<< "h" << node->height << std::endl;
	this->traverse(node->right, level+1);
}

} // namespace avl

// only here for testing while we implement the gtest module
int main()
{
	std::vector<int> test = {5,7,0,9,3,2,8};
	auto tree = avl::AVL<int>(test);
	tree.traverse();
	auto to_erase = 5;
	std::cout << "----- erasing: " << to_erase << " -----" << std::endl;
	//tree.erase(to_erase);
	//tree.traverse();
	auto node = tree.search(9);
	if(node)
		std::cout << "node found: " << std::to_string(node->value) << std::endl;
	node = tree.max();
	if(node)
		std::cout << "max node found: " << std::to_string(node->value) << std::endl;
	node = tree.min();
	if(node)
		std::cout << "min node found: " << std::to_string(node->value) << std::endl;
	node = tree.successor(8);
	if(node)
		std::cout << "successor node found: " << std::to_string(node->value) << std::endl;
	return 0;
}
